### PR TITLE
Support more ClickHouse types in ORC/Arrow/Parquet formats

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support more ClickHouse types in ORC/Arrow/Parquet formats: Enum(8|16), (U)Int(128|256), Decimal256 (for ORC), allow reading IPv4 from Int32 values (ORC outputs IPv4 as Int32 and we couldn't read it back), fix reading Nullable(IPv6) from binary data for ORC.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp [hunk 1]
```diff
@@ -537,22 +603,48 @@ static ColumnWithTypeAndName readIPv6ColumnFromBinaryData(std::shared_ptr<arrow:
...
+        }
+    }
+    return {std::move(internal_column), std::move(internal_type), column_name};
+}
+
+static ColumnWithTypeAndName readIPv4ColumnWithInt32Data(std::shared_ptr<arrow::ChunkedArray> & arrow_column, const String & column_name)
+{
+    auto internal_type = std::make_shared<DataTypeIPv4>();
+    auto internal_column = internal_type->createColumn();
+    auto & column_data = static_cast<ColumnIPv4 &>(*internal_column).getData();
```
Comment (kssenii) [inline src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:635]: `assert_cast`?

File: src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp [hunk 2]
```diff
@@ -537,22 +603,48 @@ static ColumnWithTypeAndName readIPv6ColumnFromBinaryData(std::shared_ptr<arrow:
...
+    column_data.reserve(arrow_column->length());
+
+    for (int chunk_i = 0, num_chunks = arrow_column->num_chunks(); chunk_i < num_chunks; ++chunk_i)
```
Comment (kssenii) [inline src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:638]: ?
```suggestion
    for (size_t chunk_i = 0, num_chunks = arrow_column->num_chunks(); chunk_i < num_chunks; ++chunk_i)
```

File: src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp [hunk 3]
```diff
@@ -697,6 +576,158 @@ namespace DB
...
+        const String & format_name,
+        arrow::ArrayBuilder* array_builder,
+        size_t start,
+        size_t end)
+    {
+        const auto & internal_column = assert_cast<const ColumnType &>(*write_column);
+        const auto & internal_data = internal_column.getData();
+        size_t fixed_length = sizeof(typename ColumnType::ValueType);
+        arrow::FixedSizeBinaryBuilder & builder = assert_cast<arrow::FixedSizeBinaryBuilder &>(*array_builder);
+        arrow::Status status;
```
Comment (kssenii) [inline src/Processors/Formats/Impl/CHColumnToArrowColumn.cpp:592]: why not just remove declaration and created where it is assigned? I do not see it being used before being assigned in line 598

File: src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp (hunk 2 above)
Comment (Avogar) [reply src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:638]: Index `chunk_i` is used in `arrow_column->chunk()` method and it takes `int` argument. We can use `size_t` in loop, but we will need to cast it to `int` explicitly anyway.

Comment (kssenii) [reply src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp:638]: ah, ok, did not notice, then leave as is.

**Review by kssenii — APPROVED** [review]

**Pull request merged**